src/spike/processor.js

```diff
@@ -0,0 +1,10 @@
+function defaultTransformer (key, value) {
+  return [key, value]
+}
+
+module.exports = function processor (request, field, extractors, transformers) {
+  const extractor = extractors[field]
```

Comment:
We may want a default extractor here. Does it simply look for a corresponding header?

Reply:
Maybe just debug logging is sufficient, if nothing is extracted the field is removed.

Reply:
I was thinking about that, in the end I settled on not transforming the casing because I wanted it to persist the original header.

The way I was thinking about it was if I added a header `X-Content-Test`, I know thats the key in the object, so it makes sense to look it up using: `headers['X-Content-Test']`. I feel like theres just a _littttle_ bit more cognitive overhead to be like, oh, everything is lowercased for me, I need to look it up like `headers[x-content-test]`. We could remove that overhead and lowercase all keys before we do the lookup under the hood. This would mean that we wouldn't necessarily be able to easily re-map headers to signature fields. (How do I go back from x-my-header to  X-My-Header)

Whats your take?

Reply:
We should try to have our defaults work regardless of case.

We could change the API for extractors and transformers to be functions that take all their arguments upfront and return the expected value.

```js
// extractor.js
const extractors = {
'(request-target)': (req) => {
    const { method, url } = req
    const { pathname, search } = urlParseLax(url)
    return [method.toLowerCase(), pathname, search].join(SPACE).trim()
  },
  '(created)': (req) => parseInt(req.headers.Date),
  '(expires)': (req) => parseInt(req.headers.Date) + EXPIRATION_OFFSET,
  digest: (req) => req.body,
  host: ({ url }) => {
    const { host } = urlParseLax(url)
    return host
  }
}

// This covers content-type, content-length, date, and even some custom
// signatureFields a user may require (no need to pass in their own extractor in simpler use cases).
const fallbackExtractor = (field, req) => req.headers[field]

// our default extractor
function extract (field, req) {
  const extractFn = extractors[field.toLowerCase()] || fallbackExtractor
  return extractFn(field, req)
}

module.exports = extract
```
```js
// transformer.js
const transformers = {
  digest: (key, value) => [
    key.toLowerCase(),
    [
      SHA_512_PREFIX,
      createHash(SHA_512).update(value).digest(BASE_64).toString()
    ].join('=')
  ]
}

// our default transformer
function transform (field, value) {
  const transformFn = transformers[field.toLowerCase()] || fallbackTransformer
  return transformers(field, value)
}

module.exports = transform
```

Reply:
We should try to have our defaults work regardless of case.

We could change the API for extractors and transformers to be functions that take all their arguments upfront and return the expected value.

```js
// extractor.js

const getHeader = (field, headers) => (
  Object.keys(headers).find(key => key.toLowerCase() === field.toLowerCase())
)

const extractors = {
'(request-target)': (field, req) => {
    const { method, url } = req
    const { pathname, search } = urlParseLax(url)
    return [method.toLowerCase(), pathname, search].join(SPACE).trim()
  },
  '(created)': (field, req) => parseInt(getHeader('date')),
  '(expires)': (field, req) => parseInt(getHeader('date')) + EXPIRATION_OFFSET,
  digest: (field, req) => req.body,
  host: (field, { url }) => {
    const { host } = urlParseLax(url)
    return host
  }
}

// This covers content-type, content-length, date, and even some custom
// signatureFields a user may require (no need to pass in their own extractor in simpler use cases).
const fallbackExtractor = (field, req) => req.headers[field]

// our default extractor
function extract (field, req) {
  const extractFn = extractors[field.toLowerCase()] || fallbackExtractor
  return extractFn(field, req)
}

module.exports = extract
```
```js
// transformer.js
const transformers = {
  digest: (key, value) => [
    key.toLowerCase(),
    [
      SHA_512_PREFIX,
      createHash(SHA_512).update(value).digest(BASE_64).toString()
    ].join('=')
  ]
}

// our default transformer
function transform (field, value) {
  const transformFn = transformers[field.toLowerCase()] || fallbackTransformer
  return transformers(field, value)
}

module.exports = transform
```